### PR TITLE
fix(networking): inject cluster-id-bearing GCE tag so GKE auto-firewall rules cover karpenter nodes

### LINKED ISSUE #315: bug: kubectl logs/exec/top broken on karpenter-provisioned nodes (missing cluster-id GCE tag)
**What happened**:

After upgrading to v0.3.0, `kubectl logs`, `kubectl exec`, `kubectl port-forward`, and `kubectl top node` all time out against pods running on Karpenter-provisioned nodes:

```
Error from server: Get "https://<node-ip>:10250/containerLogs/...": dial tcp <node-ip>:10250: i/o timeout
```

Metrics-server cannot scrape kubelet on Karpenter nodes, so HPA and `kubectl top` are broken cluster-wide for workloads on those nodes. Nodes themselves report `Ready` and pods report `Running` — the failure is invisible until someone inspects a pod.

**What you expected to happen**:

`kubectl logs/exec/port-forward` work normally on pods scheduled on Karpenter-provisioned nodes.

**How to reproduce it (as minimally and precisely as possible)**:

1. Upgrade to v0.3.0 (or any build after PR #284 merged).
2. Provision a node via Karpenter (create a NodePool + GCENodeClass + a workload pod).
3. Once the pod is `Running`, run `kubectl logs <pod>` — it times out.

**Anything else we need to know?**:

**Root cause:** PR #284 redesigned network tag construction to derive tags from the cluster API instead of copying from the GKE node pool template. The new code emits only the generic `gke-<clusterName>-node` tag. It dropped a second tag that was previously inherited from the template: `gke-<clusterName>-<clusterID[:8]>-node`.

GKE auto-creates cluster-level firewall rules (see [GKE firewall rules docs](https://cloud.google.com/kubernetes-engine/docs/concepts/firewall-rules#cluster-fws)) that allow the control plane to reach kubelet on TCP 10250. These rules target `gke-<clusterName>-<clusterID[:8]>-node` — not the generic tag — so without it the kubelet API is unreachable from the control plane.

**Fix:** PR #314 restores this tag by passing `cluster.Id` (already fetched from the Container API) into `buildInstanceTags`.

**Environment**:

- Karpenter-provider-gcp version: v0.3.0 (regression introduced in #284)
- GKE version: all versions
- Others: affects all standard GKE clusters that rely on GKE's auto-created firewall rules (essentially all users)

### PR DESCRIPTION
**Pull request by sleterrier**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

Restores the cluster-id-bearing GCE tag (`gke-<clusterName>-<id[:8]>-node`) that pre-#284 `buildInstanceTags` got incidentally via the bootstrap node pool template.

GKE auto-creates firewall rules ([cluster firewall rules docs](https://cloud.google.com/kubernetes-engine/docs/concepts/firewall-rules#cluster-fws)) targeting this tag (`-all`, `-vms`, `-inkubelet`, `-exkubelet`). Without it, kubelet (TCP 10250) is unreachable from the control plane, and `kubectl logs/exec/port-forward` plus metrics-server scraping silently break on every karpenter-provisioned node.

**Symptom:** `kubectl logs <pod>` against pods on karpenter-provisioned nodes fails with `dial tcp <node-ip>:10250: i/o timeout`. `kubectl top node <karpenter-node>` returns "nodemetrics not found." Pods themselves report `Ready` and run normally; only the kubelet API path from the control plane is broken.

**Fix:** The cluster's `Id` (a 64-char hex string) is already fetched via `gkeProvider.GetClusterConfig(ctx)` and threaded into `buildInstance` as `clusterConfig.Id`. This change passes it one more level into `buildInstanceTags` and appends the cluster-id-bearing tag.

Changes:
- `pkg/providers/instance/instance.go`:
  - `buildInstanceTags(clusterName, clusterID, networkTags)`: accepts the cluster ID; appends `gke-<clusterName>-<clusterID[:8]>-node`. Function comment corrected (current comment claims "GKE's cluster-level firewall rules match on this tag" referring to the hash-less tag, which is inaccurate).
  - `buildInstance` caller now passes `clusterConfig.Id`.
- `pkg/providers/instance/instance_test.go`: new `TestBuildInstanceTags` table-test verifying both tags are emitted and `networkTags` are preserved and de-duped. The three older `TestBuildInstanceTags*` functions are removed since they called the prior 2-arg signature; their scenarios are now subsumed by the table-driven test. The `makeCluster` test fixture gains a placeholder `Id` so existing tests that exercise `buildInstance` continue to pass.

/fix https://github.com/cloudpilot-ai/karpenter-provider-gcp/issues/315

#### Docs and examples

- [x] `docs/` and `examples/` updated (or this PR does not affect user-facing behaviour, configuration, or APIs)

#### Does this PR introduce a user-facing change?
None

<details open><summary><h3>Greptile Summary</h3></summary>

- Restores the cluster-id-bearing GCE tag (`gke-<clusterName>-<clusterID[:8]>-node`) that was lost in #284, fixing kubelet reachability from the control plane on Karpenter-provisioned nodes (broken `kubectl logs/exec`, metrics-server).
- `buildInstanceTags` now accepts a third `clusterID` parameter and emits the id-bearing tag only when `len(clusterID) >= 8`, safely skipping it for empty/short API responses.
- Old single-scenario tests are replaced by a table-driven `TestBuildInstanceTags` covering both tags, NodeClass tag appending, de-dup, and short/empty ID edge cases.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — targeted two-file fix with appropriate guard, no regressions found.

The previous P1 (unguarded clusterID slice) was addressed in the latest commit via the len(clusterID) >= 8 guard. The logic is straightforward, tests are comprehensive, and the change is minimal in scope.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| pkg/providers/instance/instance.go | Adds clusterID parameter to buildInstanceTags, appends cluster-id-bearing GCE tag with a safe len >= 8 guard; no logic issues found. |
| pkg/providers/instance/instance_test.go | Replaces three narrow unit tests with a comprehensive table-driven test; adds 64-char placeholder Id to makeCluster fixture so existing buildInstance tests continue to compile and pass. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant BP as buildInstance
    participant BT as buildInstanceTags
    participant GCE as GCE API

    BP->>BP: "clusterConfig = gkeProvider.GetClusterConfig(ctx)"
    BP->>BT: buildInstanceTags(clusterName, clusterConfig.Id, networkTags)
    BT->>BT: "append "gke-{clusterName}-node""
    alt "len(clusterID) >= 8"
        BT->>BT: "append "gke-{clusterName}-{clusterID[:8]}-node""
    else short/empty clusterID
        BT->>BT: skip cluster-id-bearing tag
    end
    BT->>BT: append NodeClass networkTags
    BT->>BT: lo.Uniq(items)
    BT-->>BP: "*compute.Tags"
    BP->>GCE: CreateInstance(... Tags: ...)
    Note over GCE: GKE firewall rules targeting<br/>gke-{clusterName}-{id[:8]}-node<br/>now cover the new node
```
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(networking): guard cluster-id tag ag..."](https://github.com/cloudpilot-ai/karpenter-provider-gcp/commit/4c567112b027fb048713195265e1ba1a40ea1758) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31251154)</sub>

### CODE REVIEW

File: pkg/providers/instance/instance.go [hunk 1]
```diff
@@ -978,13 +978,15 @@ func (p *DefaultProvider) belongsToCluster(inst *Instance) bool {
...
-	items := make([]string, 0, 1+len(networkTags))
+// buildInstanceTags emits both the generic gke-<clusterName>-node tag and the
+// cluster-id-bearing gke-<clusterName>-<clusterID[:8]>-node tag (the latter
+// is the target of GKE's auto-created cluster firewall rules), then appends
+// any NodeClass network tags. clusterID must be the 64-char cluster.Id from
+// the Container API.
+func buildInstanceTags(clusterName, clusterID string, networkTags []v1alpha1.NetworkTag) *compute.Tags {
+	items := make([]string, 0, 2+len(networkTags))
 	items = append(items, fmt.Sprintf("gke-%s-node", clusterName))
+	items = append(items, fmt.Sprintf("gke-%s-%s-node", clusterName, clusterID[:8]))
```
[the remark this answers is not in the document]
Comment (sleterrier) [reply pkg/providers/instance/instance.go]: Fair. Fixed in [4c56711](https://github.com/cloudpilot-ai/karpenter-provider-gcp/pull/314/commits/4c567112b027fb048713195265e1ba1a40ea1758).

**Comment by dm3ch** [comment]
@greptileai please check again

**Review by dm3ch — APPROVED** [review]

**Comment by dm3ch** [comment]
### E2E Test Results

**Cluster:** `europe-west4-a` | **Commit:** `4c567112` | **Total duration:** `~11m` (suites ran in parallel; drift dominated)

| Suite | Spec | Result | Duration |
|---|---|---|---|
| Consolidation | should consolidate an empty amd64 on-demand node | ✅ pass | 4m35s |
| Drift | should replace a drifted amd64 on-demand node | ✅ pass | 6m03s |
| Drift | should replace a node when GCENodeClass metadata changes | ✅ pass | 3m05s |
| Expiration | should replace an expired amd64 on-demand node | ✅ pass | 5m57s |
| GarbageCollection | should delete an orphaned VM after its NodeClaim is force-removed | ✅ pass | 4m01s |
| Networking | should provision a node with no external IP when enableExternalIPAccess is false | ✅ pass | 2m18s |
| Provisioning | COS / amd64 / on-demand | ✅ pass | ~3m56s |
| Provisioning | COS / arm64 / on-demand | ✅ pass | ~4m00s |
| Provisioning | COS / arm64 / spot | ✅ pass | ~4m06s |
| Provisioning | COS / amd64 / spot | ✅ pass | ~4m16s |
| Provisioning | Ubuntu / amd64 / on-demand | ✅ pass | ~4m51s |
| Provisioning | Ubuntu / arm64 / on-demand | ✅ pass | ~4m52s |
| Provisioning | Ubuntu / arm64 / spot | ✅ pass | ~4m56s |
| Provisioning | Ubuntu / amd64 / spot | ✅ pass | ~4m58s |
| Repair | should replace a node whose KernelDeadlock condition has been True beyond the toleration | ✅ pass | 4m44s |

**Overall: ✅ all 15 specs pass**
